api/turing/config/config.go

```diff
@@ -426,8 +430,8 @@ type MLPConfig struct {
 }
 
 type MlflowConfig struct {
-	TrackingURL         string
-	ArtifactServiceType string
+	TrackingURL         string `validate:"required_if=ArtifactServiceType gcs ArtifactServiceType s3"`
+	ArtifactServiceType string `validate:"required,oneof=nop gcs s3"`
```

Comment:
I'm wondering what will happen if we have a different pair of `ArtifactServiceType` and `PushRegistryType`. Like, `PushRegistryType=docker` & `ArtifactServiceType=gcs`, can still it run? I assume it can’t because then we will not have the `GOOGLE_APPLICATION_CREDENTIALS` env var for the `gcloud auth activate-service-account` command in [dockerfile](https://github.com/caraml-dev/turing/blob/2f5cc7c83eeb5bf8d1c7c79771af0453e869ab60/engines/pyfunc-ensembler-service/app.Dockerfile#L18), but CMIIW. 

Should we validate this config first?

Reply:
> I'm wondering what will happen if we have a different pair of `ArtifactServiceType` and `PushRegistryType`. Like, `PushRegistryType=docker` & `ArtifactServiceType=gcs`, can still it run?

Yes it's expected that such a configuration will still run. In other words, the `ArtifactServiceType` and `PushRegistryType` are supposed to be independent from each other. The `ArtifactServiceType` only specifies the blob storage type where Merlin model artifacts are stored whereas `PushRegistryType` specifies the image registry Merlin should push a built model image to.

> I assume it can’t because then we will not have the `GOOGLE_APPLICATION_CREDENTIALS` env var for the `gcloud auth activate-service-account` command in [dockerfile](https://github.com/caraml-dev/turing/blob/2f5cc7c83eeb5bf8d1c7c79771af0453e869ab60/engines/pyfunc-ensembler-service/app.Dockerfile#L18)
Ah yes, good catch! If `PushRegistryType=docker` & `ArtifactServiceType=gcs`, then we still need to have some form of GCS credentials in the kaniko image builder (either via a service account or via those `GOOGLE_APPLICATION_CREDENTIALS` but I've removed them from this change [here](https://github.com/caraml-dev/turing/pull/394/files#diff-4fbe33f74c84401285db6174c70b566fd9be74c9fcb30ff90a20019ff21f25e3R326). Thanks a lot for spotting this! Let me fix this!

Reply:
I've refactored the code a little (just like in Merlin) in order to add the GCP credentials (via a mounted secret and env var) if `gcs` is configured as an artifact store (this is independent from whatever is configured as the push registry (docker or gcr). You can find those changes here: https://github.com/caraml-dev/turing/pull/394/files#diff-4fbe33f74c84401285db6174c70b566fd9be74c9fcb30ff90a20019ff21f25e3R382

Reply:
Thank you so much for clarifying this and making the changes! Just one last question (promise!), in [here](https://github.com/caraml-dev/turing/blob/611615f299ee1c0ab8143a415c3b833deccd7c21/api/turing/imagebuilder/imagebuilder.go#L388-L409), it will only mount the google credential if `artifactServiceType == gcs`. But if `artifactServiceType == gcs & pushRegistryType == docker`, don't we need both the gcs & docker credential, or not?

Reply:
Ah yes that's right! Nice catch again! Hahaha thanks for giving it one last check (I must've been falling asleep when implementing that change) ><" I'll update the `if-else` block and turn it into 2 `if` blocks.